File: pipeline/jobfit/rolerubric.py
"""Role rubric — the weighted axes a role's candidates are judged against.

ADR-0010 (docs/architecture/decisions/0010-need-role-slate-one-board.md) §2: a
role FREEZES ONE rubric, derived from the RoleBrief's own graded requirements and
its ``core`` facets, and every population on the slate — a person or an AI agent —
is scored against that same list of axes. :class:`RubricAxis` is the shape of one
axis; the frozen, versioned store that persists a list of them is
``app/_lib/db/role-rubrics.ts`` (``role_rubrics.axes_json``), which validates every
write against the codegen'd ``rubricAxisSchema``.

The vocabularies are ``Literal`` on purpose: they are emitted as ``z.enum`` so a
persisted axis naming an evidence source no scorer knows is an invalid row, not a
string that silently scores nothing.

:func:`derive_role_rubric` is the derivation, deterministic and keyless by
construction: the brief's grading is already structured (kind x hardness x
weight), so turning it into axes needs no provider — an LLM could only improve the
axis PROSE, never which axes exist or what they weigh. ``app/_lib/role-rubric.ts``
mirrors it for the TS callers, and both are pinned to one fixture
(``pipeline/jobfit/tests/fixtures/role_rubric_cases.json``) so they cannot drift
apart. The rules, stated once:

* Each requirement with a skill becomes a ``requirement`` axis keyed
  ``req:<skill>``. Rows naming the same skill (case/whitespace-insensitive)
  collapse into ONE axis: the stronger kind wins (must_have over nice_to_have),
  then the stronger hardness (prerequisite over learnable), and the axis keeps the
  largest weight — a brief that says a skill twice has not asked for it to count
  twice. Label and provenance come from the first row that named the skill; the
  rationale from the first row that gave one.
* Each ``core`` facet with a name and a value becomes an axis; the first statement
  of a facet name wins. ``budget_band`` is a ``cost`` axis (``cost:budget_band``);
  every other core facet is a ``facet`` axis (``facet:<name>``). ``valuable`` and
  ``context`` facets do not become axes: ADR-0010 names only ``core`` ("losing it
  changes who you hire"), and an axis nobody said was decisive would dilute the
  ones that are.
* Raw weight = the brief's weight clamped to 0.05..1 (a stated-but-zero-weighted
  requirement still counts; a non-finite weight reads as the schema default 0.5)
  x a kind factor (must_have 1, nice_to_have 0.5); a core facet carries a fixed raw
  weight of 0.5. Axis ``weight`` is the raw weight's share of the total, so a
  rubric's axes sum to 1.
* ``blocking`` is true only for must_have x prerequisite — the one grading that
  says "without this, no". Nothing else may end a candidacy on its own.
* Order is part of the contract: requirement axes first — blocking, then
  must_have, then raw weight descending, then key — followed by the facet and cost
  axes in key order.
"""

from __future__ import annotations

import math
from typing import Literal

from .models import _Base
from .rolebrief import RoleBrief

# What in the brief produced an axis. "cost" is the budget facet — split out
# because its evidence (salary band / budget) is not a conversation.
RUBRIC_AXIS_ORIGINS = ("requirement", "facet", "cost")

# What KIND of evidence scores an axis — the rows of ADR-0010 §3's table.
RUBRIC_EVIDENCE_CLASSES = ("requirement_coverage", "demonstrated_work", "conversation", "cost")

# class -> (human evidence source, agent evidence source): the per-population
# adapters of ADR-0010 §3 ("one score, two evidence adapters"). A scorer reads
# THESE names off the axis, never its own.
RUBRIC_EVIDENCE_SOURCES: dict[str, tuple[str, str]] = {
    "requirement_coverage": ("analysis", "agent_fit"),
    "demonstrated_work": ("devcase", "trial_run"),
    "conversation": ("scorecard", "mandate_exchange"),
    "cost": ("salary_band", "budget"),
}

AxisOrigin = Literal["requirement", "facet", "cost"]
AxisKind = Literal["must_have", "nice_to_have", "core"]
AxisHardness = Literal["prerequisite", "learnable", ""]
EvidenceClass = Literal["requirement_coverage", "demonstrated_work", "conversation", "cost"]
HumanEvidence = Literal["analysis", "devcase", "scorecard", "salary_band"]
AgentEvidence = Literal["agent_fit", "trial_run", "mandate_exchange", "budget"]


class RubricAxis(_Base):
    """One weighted axis of a role's rubric. ``key`` is stable across re-derivations
    of an unchanged brief, which is what lets a score recorded against version 1 be
    read back against the axis it was actually produced under."""

    key: str = ""                          # "req:<skill>" | "facet:<name>" | "cost:<name>"
    label: str = ""                        # the requestor's own words — skill or facet label
    origin: AxisOrigin = "requirement"     # RUBRIC_AXIS_ORIGINS
    kind: AxisKind = "must_have"           # must_have | nice_to_have (requirement) | core (facet)
    hardness: AxisHardness = ""            # prerequisite | learnable; "" for a facet
    weight: float = 0.0                    # share of the rubric, 0..1; a rubric's axes sum to 1
    blocking: bool = False                 # must_have x prerequisite only
    provenance: str = "inferred"           # the brief's BRIEF_PROVENANCE for the source row
    evidence_class: EvidenceClass = "requirement_coverage"  # RUBRIC_EVIDENCE_CLASSES
    human_evidence: HumanEvidence = "analysis"              # what scores it for a person
    agent_evidence: AgentEvidence = "agent_fit"             # what scores it for an AI agent
    rationale: str = ""                    # requirement rationale / facet value, verbatim


# Core facet names whose evidence is not a conversation. Everything else a core
# facet can say ("team_context", "why_now", "success_90d", ...) is judged by talking.
_FACET_EVIDENCE_CLASS = {
    "budget_band": "cost",
    "work_environment": "demonstrated_work",
    "codebase_dossier": "demonstrated_work",
}

_KIND_FACTOR = {"must_have": 1.0, "nice_to_have": 0.5}
_KIND_RANK = {"must_have": 0, "nice_to_have": 1}
_HARDNESS_RANK = {"prerequisite": 0, "learnable": 1}
_MIN_REQUIREMENT_WEIGHT = 0.05
_DEFAULT_REQUIREMENT_WEIGHT = 0.5
_CORE_FACET_RAW_WEIGHT = 0.5


def _collapse(text: str) -> str:
    return " ".join(text.split())


def _evidence(evidence_class: str) -> dict[str, str]:
    human, agent = RUBRIC_EVIDENCE_SOURCES[evidence_class]
    return {"evidence_class": evidence_class, "human_evidence": human, "agent_evidence": agent}


def derive_role_rubric(brief: RoleBrief) -> list[RubricAxis]:
    """The rubric axes a RoleBrief states. Pure: same brief, same list, every time.

    An empty list is a real answer (a brief with no graded requirement and no core
    facet), not an error — refusing to mint an empty rubric is the store's call."""

    merged: dict[str, dict] = {}
    for req in brief.requirements:
        label = _collapse(req.skill)
        if not label:
            continue
        kind = req.kind if req.kind in _KIND_FACTOR else "must_have"
        hardness = req.hardness if req.hardness in _HARDNESS_RANK else "prerequisite"
        stated = float(req.weight) if math.isfinite(req.weight) else _DEFAULT_REQUIREMENT_WEIGHT
        weight = max(_MIN_REQUIREMENT_WEIGHT, min(1.0, stated))
        key = f"req:{label.lower()}"
        seen = merged.get(key)
        if seen is None:
            merged[key] = {
                "key": key,
                "label": label,
                "kind": kind,
                "hardness": hardness,
                "weight": weight,
                "provenance": req.provenance,
                "rationale": req.rationale.strip(),
            }
            continue
        if _KIND_RANK[kind] < _KIND_RANK[seen["kind"]]:
            seen["kind"] = kind
        if _HARDNESS_RANK[hardness] < _HARDNESS_RANK[seen["hardness"]]:
            seen["hardness"] = hardness
        seen["weight"] = max(seen["weight"], weight)
        if not seen["rationale"]:
            seen["rationale"] = req.rationale.strip()

    facets: dict[str, dict] = {}
    for facet in brief.facets:
        if facet.importance != "core":
            continue
        name = _collapse(facet.key).lower() or _collapse(facet.label).lower()
        value = facet.value.strip()
        if not name or not value:
            continue
        evidence_class = _FACET_EVIDENCE_CLASS.get(name, "conversation")
        origin = "cost" if evidence_class == "cost" else "facet"
        key = f"{origin}:{name}"
        if key in facets:
            continue  # first statement of a facet wins; a repeat is not a second axis
        facets[key] = {
            "key": key,
            "label": facet.label.strip() or facet.key.strip(),
            "origin": origin,
            "provenance": facet.provenance,
            "rationale": value,
            "evidence_class": evidence_class,
        }

    requirements = sorted(
        merged.values(),
        key=lambda r: (
            0 if (r["kind"] == "must_have" and r["hardness"] == "prerequisite") else 1,
            _KIND_RANK[r["kind"]],
            -(r["weight"] * _KIND_FACTOR[r["kind"]]),
            r["key"],
        ),
    )
    facet_rows = sorted(facets.values(), key=lambda f: f["key"])

    raws = [r["weight"] * _KIND_FACTOR[r["kind"]] for r in requirements]
    raws += [_CORE_FACET_RAW_WEIGHT] * len(facet_rows)
    # Summed left to right on purpose: the TS mirror sums in the same order, so the
    # shares are bit-identical in both languages rather than merely close.
    total = 0.0
    for raw in raws:
        total += raw

    axes: list[RubricAxis] = []
    for index, r in enumerate(requirements):
        axes.append(
            RubricAxis(
                key=r["key"],
                label=r["label"],
                origin="requirement",
                kind=r["kind"],
                hardness=r["hardness"],
                weight=raws[index] / total,
                blocking=r["kind"] == "must_have" and r["hardness"] == "prerequisite",
                provenance=r["provenance"],
                rationale=r["rationale"],
                **_evidence("requirement_coverage"),
            )
        )
    offset = len(requirements)
    for index, f in enumerate(facet_rows):
        axes.append(
            RubricAxis(
                key=f["key"],
                label=f["label"],
                origin=f["origin"],
                kind="core",
                hardness="",
                weight=raws[offset + index] / total,
                blocking=False,
                provenance=f["provenance"],
                rationale=f["rationale"],
                **_evidence(f["evidence_class"]),
            )
        )
    return axes

File: app/_lib/role-rubric.ts
import type { RubricAxis } from "./schemas.generated";
import type { RoleBrief } from "./rolespec";

// The role-rubric derivation (ADR-0012 §2) for the TS callers — a line-for-line
// mirror of pipeline/jobfit/rolerubric.py::derive_role_rubric, whose module
// docstring states the rules once. Both implementations are pinned to ONE fixture
// (pipeline/jobfit/tests/fixtures/role_rubric_cases.json, read by
// role-rubric.test.ts and test_rolerubric.py) with weights compared exactly, so a
// rule changed in one language and not the other turns a suite red.
//
// Pure, DB-free and keyless: the brief's grading is already structured, so no
// provider is involved. The frozen, versioned store the result is minted into is
// app/_lib/db/role-rubrics.ts.
//
// Known edge the parity fixture does not cover: key normalisation lower-cases with
// String#toLowerCase and orders with code-unit comparison, which match Python's
// str.lower() and code-point ordering for every Latin script (Czech included) but
// not for a handful of special cases (Turkish dotted İ, astral-plane characters).

type EvidenceClass = RubricAxis["evidenceClass"];

const EVIDENCE_SOURCES: Record<EvidenceClass, [RubricAxis["humanEvidence"], RubricAxis["agentEvidence"]]> = {
  requirement_coverage: ["analysis", "agent_fit"],
  demonstrated_work: ["devcase", "trial_run"],
  conversation: ["scorecard", "mandate_exchange"],
  cost: ["salary_band", "budget"],
};

const FACET_EVIDENCE_CLASS: Record<string, EvidenceClass> = {
  budget_band: "cost",
  work_environment: "demonstrated_work",
  codebase_dossier: "demonstrated_work",
};

type Kind = "must_have" | "nice_to_have";
type Hardness = "prerequisite" | "learnable";

const KIND_FACTOR: Record<Kind, number> = { must_have: 1, nice_to_have: 0.5 };
const KIND_RANK: Record<Kind, number> = { must_have: 0, nice_to_have: 1 };
const HARDNESS_RANK: Record<Hardness, number> = { prerequisite: 0, learnable: 1 };
const MIN_REQUIREMENT_WEIGHT = 0.05;
const DEFAULT_REQUIREMENT_WEIGHT = 0.5;
const CORE_FACET_RAW_WEIGHT = 0.5;

// The BriefRequirement / BriefFacet defaults (rolebrief.py), for rows read from a
// stored blob that predates a field.
const str = (value: unknown, fallback = ""): string => (typeof value === "string" ? value : fallback);
const collapse = (text: string): string => text.split(/\s+/).filter(Boolean).join(" ");
const isKind = (value: string): value is Kind => value === "must_have" || value === "nice_to_have";
const isHardness = (value: string): value is Hardness => value === "prerequisite" || value === "learnable";
const cmp = (a: string, b: string): number => (a < b ? -1 : a > b ? 1 : 0);

function evidence(evidenceClass: EvidenceClass): Pick<RubricAxis, "evidenceClass" | "humanEvidence" | "agentEvidence"> {
  const [humanEvidence, agentEvidence] = EVIDENCE_SOURCES[evidenceClass];
  return { evidenceClass, humanEvidence, agentEvidence };
}

type MergedRequirement = {
  key: string;
  label: string;
  kind: Kind;
  hardness: Hardness;
  weight: number;
  provenance: string;
  rationale: string;
};

type CoreFacet = {
  key: string;
  label: string;
  origin: "facet" | "cost";
  provenance: string;
  rationale: string;
  evidenceClass: EvidenceClass;
};

/** The rubric axes a RoleBrief states. Pure: same brief, same list, every time. An
 *  empty list is a real answer (no graded requirement, no core facet) — refusing to
 *  mint an empty rubric is the store's call. */
export function deriveRoleRubric(brief: RoleBrief): RubricAxis[] {
  const merged = new Map<string, MergedRequirement>();
  for (const req of brief.requirements ?? []) {
    const label = collapse(str(req?.skill));
    if (!label) continue;
    const rawKind = str(req.kind, "must_have");
    const rawHardness = str(req.hardness, "prerequisite");
    const kind: Kind = isKind(rawKind) ? rawKind : "must_have";
    const hardness: Hardness = isHardness(rawHardness) ? rawHardness : "prerequisite";
    const stated = typeof req.weight === "number" && Number.isFinite(req.weight) ? req.weight : DEFAULT_REQUIREMENT_WEIGHT;
    const weight = Math.max(MIN_REQUIREMENT_WEIGHT, Math.min(1, stated));
    const key = `req:${label.toLowerCase()}`;
    const rationale = str(req.rationale).trim();
    const seen = merged.get(key);
    if (!seen) {
      merged.set(key, { key, label, kind, hardness, weight, provenance: str(req.provenance, "inferred"), rationale });
      continue;
    }
    if (KIND_RANK[kind] < KIND_RANK[seen.kind]) seen.kind = kind;
    if (HARDNESS_RANK[hardness] < HARDNESS_RANK[seen.hardness]) seen.hardness = hardness;
    seen.weight = Math.max(seen.weight, weight);
    if (!seen.rationale) seen.rationale = rationale;
  }

  const facets = new Map<string, CoreFacet>();
  for (const facet of brief.facets ?? []) {
    if (str(facet?.importance, "valuable") !== "core") continue;
    const name = collapse(str(facet.key)).toLowerCase() || collapse(str(facet.label)).toLowerCase();
    const value = str(facet.value).trim();
    if (!name || !value) continue;
    const evidenceClass = FACET_EVIDENCE_CLASS[name] ?? "conversation";
    const origin = evidenceClass === "cost" ? "cost" : "facet";
    const key = `${origin}:${name}`;
    if (facets.has(key)) continue; // first statement of a facet wins
    facets.set(key, {
      key,
      label: str(facet.label).trim() || str(facet.key).trim(),
      origin,
      provenance: str(facet.provenance, "inferred"),
      rationale: value,
      evidenceClass,
    });
  }

  const blockingRank = (r: MergedRequirement) => (r.kind === "must_have" && r.hardness === "prerequisite" ? 0 : 1);
  const rawOf = (r: MergedRequirement) => r.weight * KIND_FACTOR[r.kind];
  const requirements = [...merged.values()].sort(
    (a, b) =>
      blockingRank(a) - blockingRank(b) ||
      KIND_RANK[a.kind] - KIND_RANK[b.kind] ||
      rawOf(b) - rawOf(a) ||
      cmp(a.key, b.key)
  );
  const facetRows = [...facets.values()].sort((a, b) => cmp(a.key, b.key));

  const raws = [...requirements.map(rawOf), ...facetRows.map(() => CORE_FACET_RAW_WEIGHT)];
  // Summed left to right, as the Python side does, so the shares are bit-identical.
  let total = 0;
  for (const raw of raws) total += raw;

  const axes: RubricAxis[] = requirements.map((r, index) => ({
    key: r.key,
    label: r.label,
    origin: "requirement",
    kind: r.kind,
    hardness: r.hardness,
    weight: raws[index] / total,
    blocking: r.kind === "must_have" && r.hardness === "prerequisite",
    provenance: r.provenance,
    ...evidence("requirement_coverage"),
    rationale: r.rationale,
  }));
  const offset = requirements.length;
  facetRows.forEach((f, index) => {
    axes.push({
      key: f.key,
      label: f.label,
      origin: f.origin,
      kind: "core",
      hardness: "",
      weight: raws[offset + index] / total,
      blocking: false,
      provenance: f.provenance,
      ...evidence(f.evidenceClass),
      rationale: f.rationale,
    });
  });
  return axes;
}

// ---------------------------------------------------------------------------
// Evaluation — ONE function, both populations (ADR-0012 §3)
// ---------------------------------------------------------------------------
//
// A person and an AI agent on one role's slate are judged against the SAME frozen
// axes by the SAME function. What differs per population is only the EVIDENCE
// ADAPTER feeding each axis: the axis itself names the human source
// (`humanEvidence`: analysis / devcase / scorecard / salary_band) and the agent
// source (`agentEvidence`: agent_fit / trial_run / mandate_exchange / budget), so
// the adapter is a lookup on the axis, never a second rubric.
//
// Deliberately NOT a single fused 0-100: blocking coverage and the remaining
// weighted coverage answer different questions and are reported SEPARATELY, and
// the unmet blocking axes are NAMED rather than deducted silently.

export type EvidencePopulation = "human" | "agent";

export type AxisEvidenceSource = RubricAxis["humanEvidence"] | RubricAxis["agentEvidence"];

/** One axis's evidence as a producer recorded it: a score in [0,1] and a pointer
 *  back to the artifact that produced it. */
export type AxisEvidence = { score: number; evidenceRef?: string | null };

/** What the caller holds about ONE candidate, keyed by axis key. An axis with no
 *  entry is "not assessed", never a zero. */
export type CandidateEvidence = {
  population: EvidencePopulation;
  axes: Record<string, AxisEvidence>;
};

/** The ADR-0012 §3 basis row: every axis score carries where it came from. */
export type AxisBasis = {
  axis: string;
  label: string;
  blocking: boolean;
  weight: number;
  /** null ⇒ this candidate has no evidence for the axis yet. */
  score: number | null;
  source: AxisEvidenceSource;
  evidenceRef: string | null;
};

export type RubricEvaluation = {
  rubricVersion: number;
  population: EvidencePopulation;
  basis: AxisBasis[];
  /** Weighted mean score over the BLOCKING axes, in [0,1]; an unassessed blocking
   *  axis counts as 0 (nobody checked it, so it is not met). 0 when the rubric has
   *  no blocking axis — "nothing required" is not "everything met". */
  blockingCoverage: number;
  /** Weighted mean over the non-blocking axes that HAVE evidence, or null when
   *  none does — an absence is not a judgement. */
  otherCoverage: number | null;
  /** Blocking axes with no evidence or a score below BLOCKING_MET_THRESHOLD. */
  unmetBlocking: string[];
  /** Axes with no evidence at all — the board says "not yet assessed" for these. */
  unassessed: string[];
};

/** A blocking axis counts as met at or above this score. */
export const BLOCKING_MET_THRESHOLD = 0.5;

/** The evidence adapter: which producer feeds this axis for this population. */
export function evidenceSourceFor(axis: RubricAxis, population: EvidencePopulation): AxisEvidenceSource {
  return population === "agent" ? axis.agentEvidence : axis.humanEvidence;
}

const clamp01 = (n: number): number => (Number.isFinite(n) ? Math.max(0, Math.min(1, n)) : 0);

/** Judge one candidate against a frozen rubric version. Pure. */
export function evaluateAgainstRubric(
  rubric: { version: number; axes: readonly RubricAxis[] },
  evidence: CandidateEvidence
): RubricEvaluation {
  const basis: AxisBasis[] = rubric.axes.map((axis) => {
    const found = Object.prototype.hasOwnProperty.call(evidence.axes, axis.key) ? evidence.axes[axis.key] : undefined;
    return {
      axis: axis.key,
      label: axis.label,
      blocking: axis.blocking,
      weight: axis.weight,
      score: found ? clamp01(found.score) : null,
      source: evidenceSourceFor(axis, evidence.population),
      evidenceRef: found?.evidenceRef ?? null,
    };
  });
  const weighted = (rows: AxisBasis[]): number | null => {
    const total = rows.reduce((s, r) => s + r.weight, 0);
    return total > 0 ? rows.reduce((s, r) => s + r.weight * (r.score ?? 0), 0) / total : null;
  };
  const blocking = basis.filter((b) => b.blocking);
  return {
    rubricVersion: rubric.version,
    population: evidence.population,
    basis,
    blockingCoverage: weighted(blocking) ?? 0,
    otherCoverage: weighted(basis.filter((b) => !b.blocking && b.score !== null)),
    unmetBlocking: blocking.filter((b) => b.score === null || b.score < BLOCKING_MET_THRESHOLD).map((b) => b.axis),
    unassessed: basis.filter((b) => b.score === null).map((b) => b.axis),
  };
}

/** Same axes, same order, same weights — "would re-deriving change the standard?".
 *  A prose-only brief edit derives identical axes. */
export function sameRubricAxes(a: readonly RubricAxis[], b: readonly RubricAxis[]): boolean {
  return JSON.stringify(a) === JSON.stringify(b);
}
